runtime/parachains/src/shared.rs

```diff
@@ -114,6 +155,148 @@ impl<T: Config> Pallet<T> {
 		Self::session_index().saturating_add(SESSION_DELAY)
 	}
 
+	/// Adds a session that is no longer in the dispute window as pruneable. Subsequent block
+	/// inclusions will incrementally remove old candidates to avoid taking the performance hit of
+	/// removing all candidates at once.
+	pub(crate) fn mark_session_pruneable(session: SessionIndex) {
+		PruneableSessions::<T>::insert(session, ());
+	}
+
+	/// Prunes up to `max_candidates_to_prune` candidates from `IncludedCandidates` that belong to
+	/// non-active sessions.
+	pub(crate) fn prune_ancient_sessions(max_candidates_to_prune: usize) {
+		let mut to_prune_candidates = Vec::new();
+		let mut to_prune_blocks = Vec::new();
+
+		let mut n_candidates = 0;
+
+		let mut incomplete_session = None;
+		for session in PruneableSessions::<T>::iter_keys() {
```

Comment:
```suggestion
		for session in PruneableSessions::<T>::iter_keys().take(max_candidates_to_prune) {
```

Reply:
I haven't read yet the rest of the code but I really think this is what you wanted